### PR TITLE
refactor: extract has and warning

### PR DESCRIPTION
**Pull request by noscripter**
According to DRY, extract `has` and `warning` message to lib as well as making the failing tests pass.

### CODE REVIEW

File: checkPropTypes.js [hunk 1]
```diff
@@ -54,7 +56,7 @@ function checkPropTypes(typeSpecs, values, location, componentName, getStack) {
             var err = Error(
               (componentName || 'React class') + ': ' + location + ' type `' + typeSpecName + '` is invalid; ' +
               'it must be a function, usually from the `prop-types` package, but received `' + typeof typeSpecs[typeSpecName] + '`.' +
-              'This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.'
```
Comment (ljharb) [inline checkPropTypes.js]: i think it's much better to leave the strings inline; gzip compresses them and dead code elimination is more able to remove them.

File: checkPropTypes.js [hunk 2]
```diff
@@ -7,12 +7,14 @@
 
 'use strict';
 
+var constant = require('./lib/constant')
```
Comment (ljharb) [inline checkPropTypes.js]: all these lines are missing semicolons.
```suggestion
var constant = require('./lib/constant');
```

File: factoryWithTypeCheckers.js [hunk 3]
```diff
@@ -11,9 +11,9 @@ var ReactIs = require('react-is');
 var assign = require('object-assign');
 
 var ReactPropTypesSecret = require('./lib/ReactPropTypesSecret');
+var has = require('./lib/has')
```
Comment (ljharb) [inline factoryWithTypeCheckers.js]:
```suggestion
var has = require('./lib/has');
```

File: checkPropTypes.js [hunk 4]
```diff
@@ -7,12 +7,14 @@
...
+var warning = constant.warning
```
Comment (ljharb) [inline checkPropTypes.js]:
```suggestion
var warning = constant.warning;
```

File: checkPropTypes.js [hunk 5]
```diff
@@ -7,12 +7,14 @@
...
 var printWarning = function() {};
 
 if (process.env.NODE_ENV !== 'production') {
   var ReactPropTypesSecret = require('./lib/ReactPropTypesSecret');
   var loggedTypeFailures = {};
-  var has = Function.call.bind(Object.prototype.hasOwnProperty);
+  var has = require('./lib/has')
```
Comment (ljharb) [inline checkPropTypes.js]:
```suggestion
  var has = require('./lib/has');
```

File: __tests__/PropTypesDevelopmentStandalone-test.js [hunk 6]
```diff
@@ -121,7 +124,8 @@ describe('PropTypesDevelopmentStandalone', () => {
       );
       expect(console.error.calls.argsFor(0)[0]).toEqual(
         'Warning: Failed prop type: testComponent: prop type `foo` is invalid; ' +
-        'it must be a function, usually from the `prop-types` package, but received `undefined`.'
+        'it must be a function, usually from the `prop-types` package, but received `undefined`.' +
+        warning
```
Comment (ljharb) [inline __tests__/PropTypesDevelopmentStandalone-test.js]: please revert the test changes here; DRY should not be applied to tests, you end up in a desert.

File: __tests__/PropTypesDevelopmentStandalone-test.js [hunk 7]
```diff
@@ -9,6 +9,7 @@
 
 'use strict';
 
+
```
Comment (ljharb) [inline __tests__/PropTypesDevelopmentStandalone-test.js]:
```suggestion
```

File: checkPropTypes.js (hunk 1 above)
Comment (noscripter) [reply checkPropTypes.js]: > i think it's much better to leave the strings inline; gzip compresses them and dead code elimination is more able to remove them.

Can you provide more information about why inline strings are better for gzip compression and dead code elimination? Thanks in advance.

**Pull request merged**